Connect rerun

esteve opened the pull request:
<!--
* Keep your PR:s small and focused.
* The PR title is what ends up in the changelog, so make it descriptive!
* If applicable, add a screenshot or gif.
* Do NOT open PR:s from your `main` branch, as that makes it hard for maintainers to test and add commits to your PR.
* Remember to run `cargo fmt` and `cargo clippy`.
* Open the PR as a draft until you have self-reviewed it and it passes CI.
* When you have addressed a PR comment, mark it as resolved.

Please be patient!
-->

Connect rclrs to Rerun.

* Closes #4

jleibs commented on src/converters/traits.rs:
```diff
@@ -0,0 +1,12 @@
+use std::sync::Arc;
+
+pub trait Converter: Send + Sync {
+    fn convert(
+        &self,
+        rec: &Arc<rerun::RecordingStream>,
+        message: &mut Cursor<Vec<u8>>,
+    ) -> Result<(), Error>;
```
```suggestion
    fn convert(
        &self,
        config: &Conversion,
        rec: &Arc<rerun::RecordingStream>,
        message: &mut Cursor<Vec<u8>>,
    ) -> Result<(), Error>;
```
I believe we should pass through the conversion config here, which is how the converter would know which entity-path to route to.

jleibs commented on src/converters/scalars.rs:
```diff
@@ -0,0 +1,155 @@
+        cdr_buffer: &mut Cursor<Vec<u8>>,
+    ) -> Result<(), Error> {
+        let value = cdr::deserialize_from::<_, i8, _>(cdr_buffer, cdr::Infinite)?;
+        rec.log("scalar", &rerun::Scalar::new(value as f64))?;
```
All of these should use the entity_path from the `Conversion` as the log destination.

jleibs commented on src/converters/mod.rs:
```diff
@@ -0,0 +1,70 @@
+        ros_type: &str,
+        message: &mut Cursor<Vec<u8>>,
+    ) -> Result<(), Error> {
+        let converter = self.get(ros_type).unwrap();
```
I believe this is where we need to look up which conversion is happening from the configuration based on `(topic, frame_id)` in order to then pass that config to the converter, to know where the routing is happening.

esteve commented:
@jleibs can you have another look at this? I've added the logic for dispatching different converters based on the `(topic, frame_id)` tuple. `frame_id` is optional in the configuration and by default conversions will just take the `topic` to create subscriptions. Thanks.

esteve commented:
@jleibs I've added a pseudo-converter for `TransformStamped`, which has a header (with a `frame_id`). The converter doesn't do anything on the `frame_id`, but I added some example code that shows how to process the `field_id` that comes from the CDR blob, see https://github.com/rerun-io/rerun-ros/pull/12/files#diff-e728a4b4964ced4a55d869ea1fd0b415c7821e12c3846b044377b8f8a7ae6578R113-R119

jleibs commented:
Makes sense.

Definitely not for this PR, but in the future it would be nice to be able to do that logic, generically, for any `Stamped` variant of a message, but I recognize there is some complexity in doing that since not all messages have headers.

Does CDR let you generically pull out a header (if you know it's there?)

Alternatively maybe this is something we handle via the trait.

If the converter looked like:
```rust
    fn convert(
        &self,
        topic: &str,
        frame_id: &Option<String>,
        entity_path: &str,
        message: &mut Cursor<Vec<u8>>,
    ) -> Result<(Option<Header>, Box<dyn Loggable>), Error>;
```

Then we could actually move the logging code out out of the convert and up into `process`. This would keep the converter from needing to know anything about the log stream or entity_paths.

Approximately something like:
```rust
let (header, loggable) = converter.convert(rec, topic, frame_id, message)?;

// Find mapping from (topic, frame_id) -> entity_path
let entity_path = lookup_path_routing(topic, header);

// Use the timestamp from the header if there is one
if let Some(header) = head {
  rec.set_time_seconds("ros_time", header.timestamp);
} else {
  rec.set_time_seconds("ros_time", receive_time);
}

rec.log(entity_path, loggable);
```

esteve commented:
> Definitely not for this PR, but in the future it would be nice to be able to do that logic, generically, for any `Stamped` variant of a message, but I recognize there is some complexity in doing that since not all messages have headers.
> 
> Does CDR let you generically pull out a header (if you know it's there?)
> 
> Alternatively maybe this is something we handle via the trait.

We could try two things:

- Attempt to read the header anyway and if it fails, process the message as usual, I have to check if the `cdr` will fail or will just deserialize whatever it's fed.
- Use the message spec representation we already have to check if there's a header field. This requires the files for the message definitions to be available in the workspace.

I think either are valid approaches, I can work on it when I come back.
